### PR TITLE
MRG: Glm refactor

### PR DESCRIPTION
**Pull request by bthirion**
Addresses #11 (see comments there).

Note that refactoring of Contrast is deffered to the future.

### CODE REVIEW

**Comment by bthirion** [comment]
I think it's now worth merging. Could somebody have a look ?

**Comment by bthirion** [comment]
Could somebody take a look at it ? It is a rather important PR...

File: nistats/tests/test_glm.py [hunk 1]
```diff
@@ -51,257 +52,198 @@ def generate_fake_fmri_data(shapes, rk=3, affine=np.eye(4)):
...
-    z_image, = multi_session_model.contrast([np.eye(rk)[1]] * 2)
-    assert_equal(np.sum(z_image.get_data() == 0), 0)
-
-    # compute the mask
-    multi_session_model = FMRILinearModel(fmri_data, design_matrices,
-                                          m=0, M=.01, threshold=0.)
-    multi_session_model.fit()
-    z_image, = multi_session_model.contrast([np.eye(rk)[1]] * 2)
+    multi_session_model = FirstLevelGLM(mask=None).fit(
+        fmri_data, design_matrices)
```
Comment (MartinPerez) [inline nistats/tests/test_glm.py]: fmri_data contains the name of files, but I do not see indication on the documentation of fit that it can also take paths instead of Nimg-like objects?

Comment (lesteve) [inline nistats/tests/test_glm.py]: Filenames are Niimg-like objects, see [this](http://nilearn.github.io/manipulating_visualizing/manipulating_images.html#niimg-like-objects) for example.

Comment (MartinPerez) [inline nistats/tests/test_glm.py]: oh i see, thanks!

File: nistats/glm.py [hunk 2]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+    -------
+    labels : array of shape (n_voxels),
+        a map of values on voxels used to identify the correseponding the model
+
+    results : dictionary,
+        keys correspond to the different labels values
+        values are RegressionResults instances corresponding to the voxels
     """
+    if noise_model not in ['ar1', 'ols']:
+        raise ValueError(' %s Unknown noise model' % noise_model)
```
Comment (lesteve) [inline nistats/glm.py]: You should probably list the acceptable noise model values and print the value that the user provided.

```python
acceptable_noise_models = ['ar1', 'ols']
if noise_model not in acceptable_noise_models:
    raise ValueError(
        'Acceptable noise models are {0}. You provided 'noise_model={1}'.format(
            acceptable_noise_models, noise_model))
```

File: nistats/glm.py [hunk 3]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
-        Parameters
-        ----------
-        X : array of shape (n_time_points, n_regressors)
-           the design matrix
-        """
-        self.X = X
-        self.labels_ = None
-        self.results_ = None
+    if Y.shape[0] != X.shape[0]:
+        raise ValueError('Response and predictors are inconsistent')
```
Comment (lesteve) [inline nistats/glm.py]: Maybe the message can be improved here too. Saying they don't have the number of rows is more precise and outputting the shape for both X and Y would help too.

File: nistats/glm.py [hunk 4]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
-        # Fit the AR model acccording to current AR(1) estimates
-        if model == 'ar1':
-            self.results_ = {}
-            self.labels_ = ar1
-            # fit the model
-            for val in np.unique(self.labels_):
-                m = ARModel(self.X, val)
-                self.results_[val] = m.fit(Y[:, self.labels_ == val])
+    # compute and discretize the AR1 coefs
+    ar1 = ((ols_result.resid[1:] * ols_result.resid[:-1]).sum(0) /
```
Comment (lesteve) [inline nistats/glm.py]: I tend to find it clearer to use the parameter name i.e. axis=0 but maybe that's just me. Same thing for the line below.

File: nistats/glm.py [hunk 5]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+            model = ARModel(X, val)
+            results[val] = model.fit(Y[:, labels == val])
+    else:
+        labels = np.zeros(Y.shape[1])
+        results = {0.0: ols_result}
+    return labels, results
+
+
+def compute_contrast(labels, regression_result, con_val, contrast_type=None):
+    """ compute the specified contrast given an estimated glm
```
Comment (lesteve) [inline nistats/glm.py]: Capital for **C**ompute

File: nistats/glm.py [hunk 6]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+
+    Parameters
+    ----------
+    labels : array of shape (n_voxels),
+        a map of values on voxels used to identify the correseponding the model
```
Comment (lesteve) [inline nistats/glm.py]: Typo: corresponding
Also one 'the' too much: identify the corresponding ~~the~~ model

File: nistats/glm.py [hunk 7]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
-            labels that associate each voxel with a results key
+    bins : int, optional
+        Maximum number of discrete bins for the AR(1) coef histogram
+
+    Returns
+    -------
+    labels : array of shape (n_voxels),
+        a map of values on voxels used to identify the correseponding the model
+
+    results : dictionary,
```
Comment (lesteve) [inline nistats/glm.py]: maybe just dict?

File: nistats/glm.py [hunk 8]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+
+def compute_contrast(labels, regression_result, con_val, contrast_type=None):
+    """ compute the specified contrast given an estimated glm
+
+    Parameters
+    ----------
+    labels : array of shape (n_voxels),
+        a map of values on voxels used to identify the correseponding the model
+
+    results : dictionary,
```
Comment (lesteve) [inline nistats/glm.py]: maybe just dict?

File: nistats/glm.py [hunk 9]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+        keys correspond to the differnt labels
```
Comment (lesteve) [inline nistats/glm.py]: keys correspond**ing**
Typo: differ**e**nt

File: nistats/glm.py [hunk 10]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+        values are RegressionResults instances corresponding to the voxels
+
+    con_val : numpy.ndarray of shape (p) or (q, p)
+        where q = number of contrast vectors and p = number of regressors
+
+    contrast_type : {None, 't', 'F' or 'tmin-conjunction'}, optional
```
Comment (lesteve) [inline nistats/glm.py]: Remove the or and use a comma instead
{None, 't', 'F', 'tmin-conjunction'}

File: nistats/glm.py [hunk 11]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+    if con_val.ndim > 1:
+        dim = con_val.shape[0]
+
+    if contrast_type is None:
+        if dim == 1:
+            contrast_type = 't'
         else:
-            self.labels_ = np.zeros(Y.shape[1])
-            self.results_ = {0.0: ols_result}
+            contrast_type = 'F'
```
Comment (lesteve) [inline nistats/glm.py]: This is the kind of case where I find ternary statements to be slightly more readable. That may just be me though:

```python
contrast_type = 't' if dim == 1 else 'F'
```

File: nistats/glm.py [hunk 12]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+
+    if contrast_type not in ['t', 'F']:
+        raise ValueError('Unknown contrast type: %s' % contrast_type)
```
Comment (lesteve) [inline nistats/glm.py]: Mentioning the list of acceptable values would be helpful.

File: nistats/glm.py [hunk 13]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+
+    effect_ = np.zeros((dim, labels.size), dtype=np.float)
```
Comment (lesteve) [inline nistats/glm.py]: np.float is platform-dependent (32 vs 64 bits). Is there a reason for this?

File: nistats/glm.py [hunk 14]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+    var_ = np.zeros((dim, dim, labels.size), dtype=np.float)
+    if contrast_type == 't':
+        for label_ in regression_result.keys():
```
Comment (lesteve) [inline nistats/glm.py]: You don't need the .keys(). Dict are iterated over keys.

File: nistats/glm.py [hunk 15]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+            resl = regression_result[label_].Tcontrast(con_val)
+            effect_[:, labels == label_] = resl.effect.T
+            var_[:, :, labels == label_] = (resl.sd ** 2).T
```
Comment (lesteve) [inline nistats/glm.py]: You recompute the mask twice. You could define a named variable for it.

File: nistats/glm.py [hunk 16]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+    dof_ = regression_result[label_].df_resid
+    return Contrast(effect=effect_, variance=var_, dof=dof_,
+                    contrast_type=contrast_type)
+
+
+class FirstLevelGLM(BaseEstimator, TransformerMixin, CacheMixin):
+    """ Implementation of the General Linear Model for Single-session fMRI data
+
+    mask: Niimg-like object, instance of NiftiMasker or MultiNiftiMasker,
+     optional
```
Comment (lesteve) [inline nistats/glm.py]: Maybe be you can do something like `Niimg-like, NiftiMasker or MultiNiftiMasker object, optional` in order to fit it on one line.

If that isn't enough you need a `\` in order for the documentation to render properly. Something like this:

```
    mask: Niimg-like object, instance of NiftiMasker or MultiNiftiMasker,\
          optional
```

File: nistats/glm.py [hunk 17]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+        Mask to be used on data. If an instance of masker is passed,
+        then its mask will be used. If no mask is given,
+        it will be computed automatically by a MultiNiftiMasker with default
+        parameters.
+
+    target_affine: 3x3 or 4x4 matrix, optional
+        This parameter is passed to image.resample_img. Please see the
```
Comment (lesteve) [inline nistats/glm.py]: maybe nilearn.image.resample_img for the sake of explicitness?

File: nistats/glm.py [hunk 18]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+        related documentation for details.
+
+    target_shape: 3-tuple of integers, optional
+        This parameter is passed to image.resample_img. Please see the
+        related documentation for details.
+
+    low_pass: False or float, optional
+        This parameter is passed to signal.clean. Please see the related
```
Comment (lesteve) [inline nistats/glm.py]: Maybe nilearn.signal.clean? A few more occurences in the neighbouring lines too.

File: nistats/glm.py [hunk 19]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+
+    noise_model : {'ar1', 'ols'}, optional
+        the temporal variance model. Defaults to 'ar1'
+
+    Attributes
+    ----------
+    labels : array of shape (n_voxels),
+        a map of values on voxels used to identify the correseponding the model
+
+    results : dictionary,
```
Comment (lesteve) [inline nistats/glm.py]: dict?

File: nistats/glm.py [hunk 20]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+        Rough estimator of the amount of memory used by caching. Higher value
+        means more memory for caching.
+
+    standardize : boolean, optional
+        If standardize is True, the time-series are centered and normed:
+        their variance is put to 1 in the time dimension.
+
+    percent_signal_change: bool, optional,
+        If True, fMRI signals is scaled to percent of the mean value
+        Incompatible with standardize (overrides standardize).
```
Comment (lesteve) [inline nistats/glm.py]: Maybe mention how standardize is set when percent_signal_change is True.

File: nistats/glm.py [hunk 21]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+        self.target_affine = target_affine
+        self.target_shape = target_shape
+        self.smoothing_fwhm = smoothing_fwhm
+        self.noise_model = noise_model
+        self.percent_signal_change = percent_signal_change
+        if self.percent_signal_change:
+            self.standardize = False
+
+    def fit(self, imgs, design_matrices):
+        """ Note: design_matrices is the design matrix !
```
Comment (lesteve) [inline nistats/glm.py]: Should this go in a note section? At the very least, there is a missing short description.

File: nistats/glm.py [hunk 22]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+                our_param = getattr(self, param_name)
+                if our_param is None:
+                    continue
+                if getattr(self.masker_, param_name) is not None:
+                    warn('Parameter %s of the masker overriden' % param_name)
+                setattr(self.masker_, param_name, our_param)
+
+        # make design_matrices a list of arrays
+        design_matrices_ = [X for X in design_matrices]
+        if isinstance(design_matrices, (_basestring, np.ndarray)):
```
Comment (lesteve) [inline nistats/glm.py]: This check is a bit weird since you just defined design_matrices to be a list in the previous line...

Comment (lesteve) [inline nistats/glm.py]: Oops not really I got confused between design_matrices and design_matrices_ (with a final underscore). Still a bit weird, should it rather be:

```python
if isinstance(design_matrices, (_basestring, np.ndarray)):
    design_matrices_ = [design_matrices]
else:
    design_matrices_ = [X for X in design_matrices]
```
```

File: nistats/glm.py [hunk 23]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
-        mse: array of shape (n_voxels)
-            the sum of square error per voxel
         """
-        # build the beta array
-        mse = np.zeros(self.labels_.size, dtype=np.float)
-        for l in self.results_.keys():
-            mse[self.labels_ == l] = self.results_[l].MSE
-        return mse
+        # First, learn the mask
+        if not isinstance(self.mask, NiftiMasker):
```
Comment (lesteve) [inline nistats/glm.py]: @arthurmensch it is not what your check_embedded_masker or something was doing ? I guess it is only in nilearn master right?

File: nistats/glm.py [hunk 24]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+            output.get_header()['descrip'] = (
+                '%s of contrast %s' % (descrip, contrast_name))
+            output_images.append(output)
+        return output_images
+
+    def fit_transform(
+        self, design_matrices, fmri_images, con_vals, contrast_type=None,
+        contrast_name='', output_z=True, output_stat=False,
+        output_effects=False, output_variance=False):
+        """ Fit then transform"""
```
Comment (lesteve) [inline nistats/glm.py]: Maybe add something like "see FirstLevelGLM.fit and FirstLevelGLM.transform documentation for more details."

File: nistats/glm.py [hunk 25]
```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
...
+
+    effect_ = np.zeros((dim, labels.size), dtype=np.float)
+    var_ = np.zeros((dim, dim, labels.size), dtype=np.float)
+    if contrast_type == 't':
+        for label_ in regression_result.keys():
+            resl = regression_result[label_].Tcontrast(con_val)
+            effect_[:, labels == label_] = resl.effect.T
+            var_[:, :, labels == label_] = (resl.sd ** 2).T
+    else:
+        for label_ in regression_result.keys():
```
Comment (lesteve) [inline nistats/glm.py]: same thing as above, no need for .keys()

File: nistats/tests/test_glm.py (hunk 1 above)
Comment (bthirion) [inline nistats/tests/test_glm.py]: We use :
'Niimg-like object or list of Niimg-like objects,
      See 
http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.'
Which is the Nilearn convention. Do you think that it may not be 
explicit enough ?

File: nistats/glm.py (hunk 15 above)
Comment (bthirion) [inline nistats/glm.py]: Sorry, I don't get what you mean.

Comment (lesteve) [inline nistats/glm.py]: you could do something like:

```python
label_mask = labels == label_

effect[:, label_mask] = res1.effect.T
var_[:, :, label_mask] = (res1.sd ** 2).T
```

Comment (bthirion) [inline nistats/glm.py]: Thanks, that's obvious indeed.

File: nistats/glm.py [hunk 26]
```diff
@@ -58,202 +47,414 @@ def data_scaling(Y):
...
+    """ Implementation of the General Linear Model for Single-session fMRI data
+
+    mask: Niimg-like, NiftiMasker or MultiNiftiMasker object, optional,
+        Mask to be used on data. If an instance of masker is passed,
+        then its mask will be used. If no mask is given,
+        it will be computed automatically by a MultiNiftiMasker with default
+        parameters.
+
+    target_affine: 3x3 or 4x4 matrix, optional
+        This parameter is passed to nilear.image.resample_img. Please see the
```
Comment (lesteve) [inline nistats/glm.py]: missing n at the end of nilearn

File: nistats/glm.py (hunk 23 above)
Comment (bthirion) [inline nistats/glm.py]: Will do it when the next Nilearn versions is released.

**Comment by bthirion** [comment]
I thinks I have addressed all issues raised do far. Any more comments ?
I'll merge next week-end if there is no further discussion.

**Comment by bthirion** [comment]
Merging. Thx for comments.
